types/lodash/index.d.ts

```diff
@@ -16602,7 +16621,8 @@ declare namespace _ {
 
     type ValueIteratee<T> = ((value: T) => any) | string | [string, any] | PartialDeep<T>;
     type ValueKeyIteratee<T> = ((value: T, key: string) => any) | string | [string, any] | PartialDeep<T>;
-    type Comparator<T> = (a: T, b: T) => boolean;
+    type Comparator<T> = (a: T, b: T) => any;
+    type Comparator2<T1, T2> = (a: T1, b: T2) => any;
```

Comment:
Why the return type is `any`?

Reply:
Lodash looks for any truthy or falsy value in the return value, so we shouldn't restrict it to just booleans. For example, see this comparator: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/20795/files#diff-6a82a43b7b79f07644334583c1add52aR671. `boolean | undefined` should be an allowed return type because lodash will treat `undefined` as a falsy value.

Reply:
I don't buy that logic -- if something will be treated as a boolean it should be typed as a `boolean`. `Math.abs("-2")` will work at runtime but isn't allowed by TypeScript.

Reply:
Returning a non-boolean value in callbacks is a common use case that people have opened issues for in the past - see #18956. If we force a boolean return value in comparators, then for consistency I think we should also force boolean return values in functions like `filter`, `some`, and `every`. But that would break #18956.

Reply:
Let's break it. TypeScript's philosophy in regard to this kind of forgiveness and implicit conversions seems to be clear and simple.

Reply:
What if we allow the return type to be `boolean | null | undefined`? That would support the use case in #18956, but also prevent the user from returning bad things like strings, objects, or functions.

Reply:
Then again, maybe it's not a big deal to force users to put `!!` in front of anything that would otherwise be undefined.

Reply:
`boolean | null | undefined` looks like a good trade-off. In particular, there will still be an error if the user forgets to `return` a value (`x => x` vs `x => { x }`) due to the fact that TS makes a distinction between `undefined` and `void`.

In Java and, even the condition in `if` statement must be of boolean type.

Reply:
In Java and the like, even conditions in `if` statements must be of boolean type, which was considered a step forward to type safety. Everyone seems to have been okay with this design decision.

Reply:
This made me think, what if change the type of `ValueIteratee` for functions like `differenceBy` from the current shape `(value: T) => any` to `(value: T) => {} | undefined | null`. If I'm not mistaken, `{} | undefined | null` means "`any` except `void`". And an iteratee callback that never returns a value is definitely a mistake.